tower-grpc-interop/src/client.rs

```diff
@@ -216,6 +216,302 @@ fn assert_success(result: Result<Vec<TestAssertion>, Box<Error>>)
+            future::ok::<Vec<TestAssertion>, Box<Error>>(assertions)
+        }))
+}
+
+fn server_streaming_test(client: &mut TestClient, core: &mut tokio_core::reactor::Core)
+        -> Result<Vec<TestAssertion>, Box<Error>> {
+    use pb::ResponseParameters;
+    let req = pb::StreamingOutputCallRequest {
+        response_parameters: RESPONSE_LENGTHS.iter().map(|len| {
+            ResponseParameters::with_size(*len)
+        }).collect(),
+        ..Default::default()
+    };
+    let req = Request::new(req);
+    core.run(client
+        .streaming_output_call(req)
+        .map_err(|tower_error| -> Box<Error> {
+            Box::new(tower_error)
+        })
+        .and_then(|response_stream| {
+            // Convert the stream into a plain Vec
+            response_stream.into_inner()
+                .collect()
+                .map_err(|tower_error| -> Box<Error> {
+                    Box::new(tower_error)
+                })
+        })
+        .map(|responses: Vec<pb::StreamingOutputCallResponse>| -> Vec<TestAssertion> {
+            let actual_response_lengths = response_lengths(&responses);
+            vec![
+                test_assert!(
+                    "there should be four responses",
+                    responses.len() == 4,
+                    format!("responses.len()={:?}", responses.len())
+                ),
+                test_assert!(
+                    "the response payload sizes should match input",
+                    RESPONSE_LENGTHS == actual_response_lengths.as_slice(),
+                    format!("{:?}={:?}", RESPONSE_LENGTHS, actual_response_lengths)
+                ),
+            ]
+        })
+        .then(&assert_success))
+}
+
+fn make_ping_pong_request(idx: usize) -> pb::StreamingOutputCallRequest {
+    let req_len = REQUEST_LENGTHS[idx];
+    let resp_len = RESPONSE_LENGTHS[idx];
+    pb::StreamingOutputCallRequest {
+        response_parameters: vec![
+            pb::ResponseParameters::with_size(resp_len)
+        ],
+        payload: Some(util::client_payload(req_len as usize)),
+        ..Default::default()
+    }
+}
+
+/// Contains the state that is threaded through the recursive "loop"
+/// of futures that performs the actual ping-pong with the server.
+struct PingPongState {
+    sender: futures::sync::mpsc::UnboundedSender<pb::StreamingOutputCallRequest>,
+    stream: Box<Stream<Item=pb::StreamingOutputCallResponse, Error=tower_grpc::Error>>,
+    responses: Vec<pb::StreamingOutputCallResponse>,
+    assertions: Vec<TestAssertion>,
+}
+
+type PingPongResponsesFuture = Box<Future<
+    Item=(Vec<pb::StreamingOutputCallResponse>, Vec<TestAssertion>),
+    Error=Box<Error>>>;
+/// Recursive function that takes the current PingPongState of the ping-pong
+/// operation and performs the rest of it. It returns a future of
+/// responses from the server and assertions that were performed in
+/// the process.
+fn perform_ping_pong(state: PingPongState) -> PingPongResponsesFuture {
```

Comment:
Nit: this seems like it _could_ be a method on `PingPongState`...

Reply:
Done.